### PR TITLE
EZP-25009: Role assignment with  limitations

### PR DESCRIPTION
**Pull request by StephaneDiot**
jira: https://jira.ez.no/browse/EZP-25009

## Description

The goal is to be able to assign a role to a user or user group with a section and subtree limitation.

:warning: the redirection is not implemented after the role is assigned. Seems it will wait for https://jira.ez.no/browse/EZP-25442

:warning:² <s>Currently it is working with IDs instead of URIs for REST request to assign role to user with limitation. It's not what we are expecting from the REST API. This will be changed when https://github.com/ezsystems/ezpublish-kernel/pull/1569 will be finnish. So it shouldn't be merged before.</s>  Now it's done !
## Tests

- Manual tests
- Unit tests

## TODO
- [ ] replace IDs by URIs

### CODE REVIEW

**Comment by StephaneDiot** [comment]
ready for early review @dpobel

**Comment by StephaneDiot** [comment]
also @yannickroger

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 1]
```diff
@@ -34,6 +37,34 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                    limitationType: 'Section',
+                    sectionId: sectionSelector.get('options').item(sectionSelectedIndex).get('value'),
+                    sectionName: sectionSelector.get('options').item(sectionSelectedIndex).get('text'),
+                };
+
+            e.preventDefault();
+            this._uiSetAssignRoleLoading(button);
+            this._fireContentDiscover(button, unsetLoading, udwConfigData);
+        },
+        /**
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: nitpick: missing empty line

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 2]
```diff
@@ -41,20 +72,34 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+        },
+
+        /**
+         * Fire contentDiscover event to launch the UDW with a config using the given data
+         *
+         * @method _fireContentDiscover
+         * @protected
+         * @param {Y.Node} button
+         * @param {Y.Function} unsetLoading
+         * @param {Y.Object} data
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: `Y.Object` ? no just `Object`

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 3]
```diff
@@ -34,6 +37,34 @@ YUI.add('ez-roleserversideview', function (Y) {
         /**
          * tap event handler on the role assign buttons. It launches the
          * universal discovery widget so that the user can pick some contents.
+         * It also define the section limitation
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: no sure what that means ?

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 4]
```diff
@@ -38,10 +38,15 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
             var data = e.target.get('data'),
                 userService = this.get('capi').getUserService(),
                 countAssigned = 0,
-                service = this;
+                service = this,
+                limitation = null;
+
+            if (this._hasSectionLimitation(data)) {
+               limitation = this._createSectionLimitation(e);
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: I would not pass the event facade in parameter

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 5]
```diff
@@ -187,14 +223,18 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
          * @param {Integer} countAssigned number of successfully assignments
          * @param {Function} callback the callback to call when other tasks are done
          */
-        _assignRoleCallback: function (roleId, roleName, contents, countAssigned, callback) {
+        _assignRoleCallback: function (data, contents, countAssigned, callback) {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: if you change the parameter, you need to change the JSDoc block. Also `data` is a very generic name, you should try to find something a bit more precise.

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 6]
```diff
@@ -216,17 +256,20 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
          *
          * @method _assignRoleNotificationStarted
          * @protected
-         * @param {String} roleId the role id
-         * @param {String} roleName the role name
+         * @param {Y.Object} data it contains the role id and role name
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: `Object` not `Y.Object`

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 7]
```diff
@@ -71,18 +76,53 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
             });
         },
 
+
+        /**
+         * Check if the UDW config has a section limitationType
+         *
+         * @method _hasSectionLimitation
+         * @protected
+         * @param {Y.Object} data
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: `Object` and missing `@return`

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 8]
```diff
@@ -238,17 +281,16 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
          *
          * @method _loadRoleErrorNotification
          * @protected
-         * @param {String} roleId the role id
-         * @param {String} roleName the role name
+         * @param {Y.Object} data it contains the role name
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: `Object` (same several times below)

**Comment by StephaneDiot** [comment]
ready for review @dpobel @yannickroger

File: Resources/views/Role/assign_role_by_section.html.twig [hunk 9]
```diff
@@ -26,7 +26,8 @@
         <select id="data-role-assignment-section" class="ez-role-assignment-section-id">
             {% for section in sections %}
             {# @var section \eZ\Publish\API\Repository\Values\Content\Section #}
-            <option value="{{ section.id }}">{{ section.name }}</option>
+                <option data-section-rest-id="{{ path('ezpublish_rest_loadSection', {'sectionId': section.id}) }}"
+                <option value="{{ section.id }}">{{ section.name }}</option>
```
Comment (dpobel) [inline Resources/views/Role/assign_role_by_section.html.twig]: 2 `<option>` opening tag ?

File: Resources/public/js/apps/ez-platformuiapp.js [hunk 10]
```diff
@@ -526,8 +526,8 @@ YUI.add('ez-platformuiapp', function (Y) {
             this.get('container').addClass(viewInfo.hideClass);
             if ( viewInfo.instance ) {
                 view = viewInfo.instance;
-                view.set('active', false);
                 viewInfo.serviceInstance.removeTarget(this);
+                view.set('active', false);
```
Comment (dpobel) [inline Resources/public/js/apps/ez-platformuiapp.js]: is this change still necessary? to be honest, I'm a bit afraid of the side effects ?

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 11]
```diff
@@ -12,10 +12,16 @@ YUI.add('ez-roleserversideview', function (Y) {
     Y.namespace('eZ');
 
     var events = {
-            '.ez-role-assign-button': {
-                'tap': '_pickSubtree'
-            },
-        };
+        '.ez-role-assign-button': {
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: nitpick: indentation issue

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 12]
```diff
@@ -34,6 +40,36 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                container = this.get('container'),
+                sectionSelector = container.one(".ez-role-assignment-section-id"),
+                sectionSelectedIndex = sectionSelector.get('selectedIndex'),
+                unsetLoading = Y.bind(this._uiUnsetAssignRoleLoading, this, button),
+                udwConfigData = {
+                    roleId: button.getAttribute('data-role-rest-id'),
+                    roleName: button.getAttribute('data-role-name'),
+                    afterUpdateCallback: unsetLoading,
+                    limitationType: 'Section',
+                    sectionId: sectionSelector.get('options').item(sectionSelectedIndex).get('value'),
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: I would built a plain object representing the section

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 13]
```diff
@@ -41,20 +77,98 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+         * @param {EventFacade} e
+         */
+        _setSubtreeLimitation: function (button, udView, e) {
+            var unsetLoading = Y.bind(this._uiUnsetAssignRoleLoading, this, button),
+                selectedLocationsIds = [],
+                that = this,
+                udwConfigData,
+                udwAfterActiveChangeEvent;
+
+            Y.Array.each(e.selection, function (struct) {
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: using `Y.Array.map` would be a tad more elegant

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 14]
```diff
@@ -41,20 +77,98 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+                limitationType: 'Subtree',
+                subtreeIds: selectedLocationsIds,
+            };
+
+
+            udwAfterActiveChangeEvent = udView.onceAfter('activeChange', function() {
+                udwAfterActiveChangeEvent.detach();
+                    setTimeout(function() {
+                        that._fireContentDiscover(button, unsetLoading, udwConfigData);
+                    },100);
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: do you need to set an explicit timeout ?

File: Resources/public/js/views/serverside/ez-roleserversideview.js [hunk 15]
```diff
@@ -41,20 +77,98 @@ YUI.add('ez-roleserversideview', function (Y) {
...
+
+            udwConfigData = {
+                roleId: button.getAttribute('data-role-rest-id'),
+                roleName: button.getAttribute('data-role-name'),
+                afterUpdateCallback: unsetLoading,
+                limitationType: 'Subtree',
+                subtreeIds: selectedLocationsIds,
+            };
+
+
```
Comment (dpobel) [inline Resources/public/js/views/serverside/ez-roleserversideview.js]: nitpick: too many lines

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 16]
```diff
@@ -71,18 +93,90 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
+         * @protected
+         * @param {Object} data
+         * @return {Object} the limitation with its identifer and values as expected by the API
+         */
+        _createSubtreeLimitation: function (data) {
+            // limitation will be filled like example in UserService.prototype.newRoleAssignInputStruct()
+            var limitation,
+                limitationValues =  [];
+
+            Y.Array.each(data.subtreeIds, function (subtreeId) {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: `Y.Array.map` here as well

File: Resources/public/js/views/services/ez-roleserversideviewservice.js [hunk 17]
```diff
@@ -71,18 +93,90 @@ YUI.add('ez-roleserversideviewservice', function (Y) {
...
+                limitationValues.push({"_href": subtreeId, "_media-type": "application\/vnd.ez.api.Location+json"});
+            });
+                limitation = {
```
Comment (dpobel) [inline Resources/public/js/views/services/ez-roleserversideviewservice.js]: nitpick: indentation issue here as well

File: Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js [hunk 18]
```diff
@@ -175,12 +171,32 @@ YUI.add('ez-roleserversideviewservice-tests', function (Y) {
                 },
                 contentType = getMockForJson(contentTypeJson),
                 selection = [{contentInfo: contentInfo, contentType: contentType, location: location}],
-                universalDiscovery = new Mock(),
+                config = {data: {roleId: '42'}};
+
+            Mock.expect(this.userService, {
+                method: 'assignRoleToUserGroup',
+                args: [Mock.Value.String, this.roleAssignInputStruct, Mock.Value.Function],
```
Comment (dpobel) [inline Tests/js/views/services/assets/ez-roleserversideviewservice-tests.js]: there's no assert on the first parameter

**Comment by StephaneDiot** [comment]
update @dpobel @yannickroger

**Comment by StephaneDiot** [comment]
Closed due to some bad git handling ... re opened : https://github.com/ezsystems/PlatformUIBundle/compare/master...StephaneDiot:ezp-25009_Role_assgignement
